fabtests/prov/efa/src/rdm_cs.c

```diff
@@ -0,0 +1,178 @@
+ * It tests the path of av remove and re-insert during the client's
+ * leaving and re-joining.
+ */
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <getopt.h>
+
+#include <shared.h>
+
+static int run_pingpong(void)
+{
+	int ret, i;
+
+	for (i = 0; i < opts.iterations; i++) {
+		if (opts.dst_addr) {
+			ret = ft_tx(ep, remote_fi_addr, opts.transfer_size, &tx_ctx);
+			if (ret) {
+				FT_PRINTERR("ft_tx", -ret);
+				return ret;
+			}
+			ret = ft_rx(ep, opts.transfer_size);
+			if (ret) {
+				FT_PRINTERR("ft_rx", -ret);
+				return ret;
+			}
+		} else {
+			ret = ft_rx(ep, opts.transfer_size);
+			if (ret) {
+				FT_PRINTERR("ft_rx", -ret);
+				return ret;
+			}
+			ret = ft_tx(ep, remote_fi_addr, opts.transfer_size, &tx_ctx);
+			if (ret) {
+				FT_PRINTERR("ft_tx", -ret);
+				return ret;
+			}
+		}
+	}
+	return 0;
+}
+
+static int run(void)
+{
+	int ret;
+	int nconn = 1;
+
+	ret = ft_init_fabric();
+	if (ret) {
+		FT_PRINTERR("ft_init_fabric", ret);
+		goto out;
+	}
+
+	if ((opts.options & FT_OPT_SERVER_PERSIST) && !opts.dst_addr)
+		nconn = opts.num_connections;
+
+	while (nconn) {
+		ret = ft_send_recv_greeting(ep);
+		if (ret) {
+			FT_PRINTERR("ft_send_recv_greeting", ret);
+			goto out;
+		}
+		ret = run_pingpong();
+		if (ret)
+			goto out;
+		if (--nconn) {
+			ret = fi_av_remove(av, &remote_fi_addr, 1, 0);
+			if (ret) {
+				FT_PRINTERR("fi_av_remove", ret);
+				goto out;
+			}
+			ret = ft_accept_next_client();
+			if (ret) {
+				FT_PRINTERR("ft_accept_next_client", ret);
+				goto out;
+			}
+		}
+	}
+out:
+	ft_free_res();
+	return ret;
+}
+
+int main(int argc, char **argv)
+{
+	int op, ret;
+	int client_count = 0;
+	char client_count_str[8];
+
+	opts = INIT_OPTS;
+	opts.options |= FT_OPT_SIZE;
+	opts.options |= FT_OPT_OOB_CTRL;
+
+	hints = fi_allocinfo();
+	if (!hints)
+		return EXIT_FAILURE;
+
+	while ((op = getopt(argc, argv, "Uh" ADDR_OPTS INFO_OPTS CS_OPTS)) != -1) {
+		switch (op) {
+		default:
+			ft_parse_addr_opts(op, optarg, &opts);
+			ft_parseinfo(op, optarg, hints, &opts);
+			ft_parsecsopts(op, optarg, &opts);
+			break;
+		case 'U':
+			hints->tx_attr->op_flags |= FI_DELIVERY_COMPLETE;
+			break;
+		case '?':
+		case 'h':
+			ft_usage(argv[0], "EFA RDM client-server test.");
+			return EXIT_FAILURE;
+		}
+	}
+
+	if (optind < argc)
+		opts.dst_addr = argv[optind];
+
+	if (opts.dst_addr) {
+		if (getenv("CLIENT_COUNT"))
+			client_count = atoi(getenv("CLIENT_COUNT"));
+	}
+
+	hints->ep_attr->type = FI_EP_RDM;
+	hints->caps = FI_MSG;
+	hints->mode = FI_CONTEXT;
+	hints->domain_attr->mr_mode = opts.mr_mode;
+	ret = run();
+
+	if (opts.dst_addr) {
+		client_count++;
+		if ((ret) || (client_count == opts.num_connections)) {
+			goto out;
+		} else {
+			sprintf(client_count_str, "%d", client_count);
+			setenv("CLIENT_COUNT", client_count_str, 1);
+			ret = ft_exec(argv);
+			if (ret) {
+				FT_PRINTERR("ft_exec", ret);
+				goto out;
+			}
+		}
+	}
+out:
+	return ft_exit_code(ret);
+}
```

Comment:
I don't see anything in this test that would make it provider specific.  Why isn't it just a new common test?